File: algorithms/java/src/TwoSum.java
import java.lang.reflect.MalformedParameterizedTypeException;
import java.util.HashMap;
import java.util.Map;

/**
 * @author jasper
 * @create 2021-09-23 18:15
 */
public class TwoSum {

    /**
     * 暴力解法
     * 时间复杂度：O(n²)
     * 空间复杂度：O(1)
     */
    public int[] twoSum1(int[] nums, int target) {
        int[] result = new int[2];
        for (int i = 0; i < nums.length; i++) {
            for (int j = (i + 1); j < nums.length; j++) {
                if (nums[i] + nums[j] == target) {
                    result[0] = i;
                    result[1] = j;
                }
            }
        }
        return result;
    }

    /**
     * 使用has会map，空间换时间
     * 时间复杂度：O(n)
     * 空间复杂度：O(n)
     */
    public int[] twoSum2(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            map.put(nums[i], i);
        }
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(target - nums[i]) && map.get(target - nums[i]) != i) {
                return new int[]{i, map.get(target - nums[i])};
            }
        }
        throw new IllegalArgumentException("ERROR");
    }
}

File: algorithms/python/TwoSum.py
# @author jasper
# @create 2021-09-23 18:39
from typing import List


class Solution:

    # 暴力解法
    # 时间复杂度：O(n²)
    # 空间复杂度：O(1)
    def twoSum1(self, nums: List[int], target: int) -> List[int]:
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i, j]

    # 使用has会map，空间换时间
    # 时间复杂度：O(n)
    # 空间复杂度：O(n)
    def twoSum2(self, nums: List[int], target: int) -> List[int]:
        d = {}
        for i in range(len(nums)):
            if target - nums[i] in d.keys():
                return [d[target - nums[i]], i]
            else:
                d[target - nums[i]] = i
